{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#chroma api #import packages\n",
    "from chroma import api\n",
    "api.register_key(\"e424a1b4a1604a3a8cc83f0792dc3253\")\n",
    "from chroma import Protein, Chroma\n",
    "import os\n",
    "from Bio import PDB\n",
    "import pandas as pd\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#file pathes\n",
    "design_out = './design_out/'\n",
    "prediction_out = './prediction_out/'\n",
    "output_file_path = \"./output.tsv\" \n",
    "tsv_file_path = \"./output.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "#esmfold prediction\n",
    "\n",
    "output_path=\"/Users/jianchengluo/Desktop/mycoplasma/prediction_out/\"\n",
    "file_path=\"protein.fna\"\n",
    "sequences=$(awk '/^[>]/ {if (seq) {print seq}; seq=\"\"; next} /^[^>]/ {gsub(/\\n/, \"\"); seq = seq $0} END {print seq}' \"$file_path\")\n",
    "\n",
    "counter=1\n",
    "while read -r sequence; do\n",
    "  sequence_length=$(echo -n \"$sequence\" | wc -c)\n",
    "  if [ \"$sequence_length\" -gt 400 ]; then\n",
    "    echo \"Skipping sequence $counter as its length is greater than 400\"\n",
    "    ((counter++))\n",
    "    continue\n",
    "  fi\n",
    "\n",
    "  pdb_file=\"$output_path$counter.pdb\"\n",
    "  while true; do\n",
    "    curl -X POST --data \"$sequence\" \"https://api.esmatlas.com/foldSequence/v1/pdb/\" > \"$pdb_file\"    \n",
    "    if [ -z \"$(cat \"$pdb_file\")\" ] || grep -q \"INTERNAL SERVER ERROR\" \"$pdb_file\"; then\n",
    "      echo \"PDB file has INTERNAL SERVER ERROR or is empty, retrying...\"\n",
    "    else\n",
    "      echo \"PDB file generated successfully: $pdb_file\"\n",
    "      break\n",
    "    fi\n",
    "  done\n",
    "  ((counter++))\n",
    "\n",
    "done <<< \"$sequences\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#design proteins\n",
    "\n",
    "chroma = Chroma()\n",
    "protein_names = [file for file in os.listdir(prediction_out) if file.endswith(\".pdb\")]\n",
    "\n",
    "for p in protein_names:\n",
    "    print(p)\n",
    "    protein = Protein(prediction_out + p)\n",
    "    protein = chroma.design(protein)\n",
    "    protein.to(design_out + p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#PDB to pro_sequence\n",
    "\n",
    "def pdb_to_sequence(pdb_file_path):\n",
    "    parser = PDB.PDBParser(QUIET=True)\n",
    "    structure = parser.get_structure('protein', pdb_file_path)\n",
    "    sequence = \"\"\n",
    "    for model in structure:\n",
    "        for chain in model:\n",
    "            for residue in chain:\n",
    "                try:\n",
    "                    amino_acid = PDB.Polypeptide.three_to_one(residue.resname)\n",
    "                except KeyError:\n",
    "                    amino_acid = 'X'\n",
    "                sequence += amino_acid    \n",
    "    return sequence\n",
    "\n",
    "df = pd.read_csv(output_file_path, delimiter='\\t')\n",
    "pdb_names = [file for file in os.listdir(design_out) if file.endswith(\".pdb\")]\n",
    "pdb_names = sorted(pdb_names, key=lambda x: int(x.split('.')[0]))\n",
    "\n",
    "df['pro_sequence'] = \"\"\n",
    "protein_sequences = []\n",
    "\n",
    "\n",
    "for pdb in pdb_names:\n",
    "    protein_sequence = pdb_to_sequence(design_out + pdb)\n",
    "    protein_sequences.append(protein_sequence)\n",
    "\n",
    "for index, protein_sequence in enumerate(protein_sequences):\n",
    "    df.at[index, 'pro_sequence'] = protein_sequence\n",
    "    print(protein_sequence)\n",
    "\n",
    "df.to_csv(output_file_path, sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#protein_sequence to DNA sequence\n",
    "\n",
    "def protein_to_dna(protein_sequence):\n",
    "    genetic_code = {\n",
    "        'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "        'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "        'N': ['AAT', 'AAC'],\n",
    "        'D': ['GAT', 'GAC'],\n",
    "        'C': ['TGT', 'TGC'],\n",
    "        'Q': ['CAA', 'CAG'],\n",
    "        'E': ['GAA', 'GAG'],\n",
    "        'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "        'H': ['CAT', 'CAC'],\n",
    "        'I': ['ATT', 'ATC', 'ATA'],\n",
    "        'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "        'K': ['AAA', 'AAG'],\n",
    "        'M': ['ATG'],\n",
    "        'F': ['TTT', 'TTC'],\n",
    "        'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "        'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "        'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "        'W': ['TGG'],\n",
    "        'Y': ['TAT', 'TAC'],\n",
    "        'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "        '*': ['TAA', 'TAG', 'TGA']\n",
    "    }\n",
    "\n",
    "    dna_sequence = \"\"\n",
    "    for amino_acid in protein_sequence:\n",
    "        codons = genetic_code.get(amino_acid, [''])\n",
    "        chosen_codon = codons[0]\n",
    "        dna_sequence += chosen_codon\n",
    "    return dna_sequence\n",
    "\n",
    "df = pd.read_csv(output_file_path, delimiter='\\t')\n",
    "protein_seqs = df['pro_sequence'].tolist()\n",
    "\n",
    "df['sequence'] = \"\"\n",
    "\n",
    "for index, protein_seq in enumerate(protein_seqs):\n",
    "    dna_seq = protein_to_dna(protein_seq)\n",
    "    df.at[index, 'sequence'] = dna_seq\n",
    "    print(f\"Protein Sequence: {protein_seq}\")\n",
    "    print(f\"DNA Sequence: {dna_seq}\")\n",
    "    print(f\"DNA Sequence written to {output_file_path}\")\n",
    "\n",
    "df.to_csv(output_file_path, sep='\\t', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#change no \\n\n",
    "with open('./genome.txt', 'r') as file:\n",
    "    content = file.read()\n",
    "\n",
    "content_without_newlines = content.replace('\\n', '')\n",
    "\n",
    "with open('./genome.txt', 'w') as file:\n",
    "    file.write(content_without_newlines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#replace the original sequence\n",
    "\n",
    "def replace_substring(original_string, target_substring, replacement_string):\n",
    "    modified_string = original_string.replace(target_substring, replacement_string)\n",
    "    return modified_string\n",
    "\n",
    "#original_string\n",
    "with open(\"./genome.txt\", \"r\", encoding=\"utf-8\") as file:\n",
    "    genome = file.read()\n",
    "original_string = genome\n",
    "\n",
    "#target_substrings\n",
    "file_path = \"original_dna_sequence_list.txt\"  \n",
    "target_substrings = []\n",
    "with open(file_path, 'r') as file:\n",
    "    for line in file:\n",
    "        if len(line) < 1200:\n",
    "            target_substrings.append(line.rstrip())\n",
    "\n",
    "\n",
    "#replacement_strings \n",
    "replacement_strings = []\n",
    "with open(tsv_file_path, newline='') as tsvfile:\n",
    "    reader = csv.reader(tsvfile, delimiter='\\t')\n",
    "    header = next(reader)\n",
    "    sequence_index = header.index('sequence')\n",
    "    for row in reader:\n",
    "        replacement_strings.append(row[sequence_index])\n",
    "\n",
    "#replacement\n",
    "new_genome = original_string\n",
    "for target_substring,replacement_string in zip(target_substrings,replacement_strings):\n",
    "    new_genome = replace_substring(new_genome, target_substring, replacement_string)\n",
    "\n",
    "with open(\"./new_genome.txt\", \"w\", encoding=\"utf-8\") as new_file:\n",
    "    new_file.write(new_genome)\n",
    "\n",
    "\n",
    "print(\"A New Life Is Born!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change 80 \\n\n",
    "with open('./new_genome.txt', 'r') as file:\n",
    "    content2 = file.read()\n",
    "content2_with_newlines = '\\n'.join([content2[i:i+80] for i in range(0, len(content2), 80)])\n",
    "\n",
    "with open('./new_genome.txt', 'w') as file:\n",
    "    file.write(content2_with_newlines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "str1 = new_genome\n",
    "str2 = genome\n",
    "print(str1 == str2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chroma",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
